### PR TITLE
Added MLP Mixer Model

### PR DESCRIPTION
**Pull request by shireenchand**
Added the mlpmixer.py file which consists of the model as well as loading the weights from hugging face

### CODE REVIEW

**Comment by shireenchand** [comment]
Hi @juliagsy, can you please review this PR 
Thanks!

File: ivy_models/mlpmixer/mlpmixer.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        y = ivy.matrix_transpose(y)
+        y = self.mlp_block1(y)
+        y = ivy.matrix_transpose(y)
+        x = x + y
+        y = self.norm2(x)
+        y = x + self.mlp_block2(y)
+        return y
+
+
+class MLPMixer(ivy.Module):
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:42]: please extend this from `BaseModel`
ref: `ivy_models/base` folder

File: ivy_models/mlpmixer/mlpmixer.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    def __init__(
+        self,
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:44]: there's some general conventions to follow, more info in this guide! 
ref: https://docs.google.com/document/d/1NfVngAQlQQlxLGEg47O2lnfn60M5xZekZFei9k5RJd8/edit?usp=sharing

File: ivy_models/mlpmixer/mlpmixer.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        num_blocks=num_blocks,
+        patch_size=patch_size,
+        hidden_dim=hidden_dim,
+        token_mlp_dim=token_mlp_dim,
+        channel_mlp_dim=channel_mlp_dim,
+    )
+    if pretrained:
+        REPO_ID = "unifyai/MLPMixer"
+        FILENAME = "weights.pickled"
+        weights = joblib.load(hf_hub_download(repo_id=REPO_ID, filename=FILENAME))
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:109]: once the `BaseModel` class is inherited, we have a built-in load method, where you can use `MLPMixer.load_from_huggingface(repo_id)` and it'll return the loaded model object, so it'll be something like 
```
if pretrained:
    return MLPMixer.load_from_huggingface(repo_id)
return model = MLPMixer(
         vol,
         inp_dim,
         num_classes=num_classes,
         num_blocks=num_blocks,
         patch_size=patch_size,
         hidden_dim=hidden_dim,
         token_mlp_dim=token_mlp_dim,
         channel_mlp_dim=channel_mlp_dim,
     )
```

File: ivy_models/mlpmixer/mlpmixer.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+import ivy
+from huggingface_hub import hf_hub_download
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:2]: this import will not be needed after the refactor

File: ivy_models/mlpmixer/mlpmixer.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import joblib
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:3]: this import will not be needed after the refactor

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey shireen! Thanks for the great work! I've left some comment related to refactor and model loading below, please feel free to lmk if you've any questions, and please request another review when it's done, thanks!

File: ivy_models/mlpmixer/mlpmixer.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                num_classes=num_classes,
+                num_blocks=num_blocks,
+                patch_size=patch_size,
+                hidden_dim=hidden_dim,
+                token_mlp_dim=token_mlp_dim,
+                channel_mlp_dim=channel_mlp_dim,
+                data_format=data_format,
+            )
+        )
+        self.num_blocks = num_blocks
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:97]: as we have the spec above, the individual attributes shouldn't be needed

File: ivy_models/mlpmixer/mlpmixer.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        self.vol = vol
+        self.token_mlp_dim = token_mlp_dim
+        self.channel_mlp_dim = channel_mlp_dim
+        self.num_classes = num_classes
+
+        super(MLPMixer, self).__init__(v=v)
+
+    def _build(self, *args, **kwargs):
+        self.conv = ivy.Conv2D(
+            self.inp_dim,
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:110]: it should then be something like `self.spec.inp_dim` here

File: ivy_models/mlpmixer/mlpmixer.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        self.norm = ivy.LayerNorm([self.hidden_dim])
+        self.dropout = ivy.Dropout(0.25)
+        self.linear = ivy.Linear(self.hidden_dim, self.num_classes)
+
+    @classmethod
+    def get_spec_class(self):
+        return MLPMixerSpec
+
+    def _forward(self, x):
+        x = self.conv(x)
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:130]: (please double check but) as we would like to support dual format, I think there should be an additional check here, something like, 
```
def _forward(self, x, data_format=None):
        data_format = data_format if data_format else self.spec.data_format
        if data_format == "NHWC":
            x = ivy.permute_dims(x, (0, 3, 1, 2))
        .....
```
please ref alexnet model for more details!

File: ivy_models/mlpmixer/mlpmixer.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        num_classes=num_classes,
+        num_blocks=num_blocks,
+        patch_size=patch_size,
+        hidden_dim=hidden_dim,
+        token_mlp_dim=token_mlp_dim,
+        channel_mlp_dim=channel_mlp_dim,
+    )
+    if pretrained:
+        return MLPMixer.load_from_huggingface(
+            repo_id="unifyai/MLPMixer", weights_path="weights.hdf5"
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:169]: I've briefly gone through the hf repo, and I think the `config.json` is missing, the required files will be uploaded if the method `push_to_huggingface` is used when uploading the trained model, could you please reupload with that? thanks a lot!

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey shireen! Thanks again for the changes! They look great and just a few more points to address before we merge it!
Btw, I think the tests are missing, will you be working on it in another pr just to make sure? Thanks!

**Comment by shireenchand** [comment]
Hi @juliagsy, I have made the changes as well as added the config.json file. I'll be adding the test file in another PR.
Do let me know if there are any more changes!

File: ivy_models/mlpmixer/mlpmixer.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+            num_blocks=num_blocks,
+            patch_size=patch_size,
+            hidden_dim=hidden_dim,
+            token_mlp_dim=token_mlp_dim,
+            channel_mlp_dim=channel_mlp_dim,
+            data_format=data_format,
+        )
+
+
+class MLPBlock(BaseModel):
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:31]: sorry, might be a misunderstanding here, only the main model class needs to be inherited from `BaseModel`, the other should just be `ivy.Module`

File: ivy_models/mlpmixer/mlpmixer.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    def _forward(self, x):
+        y = self.linear0(x)
+        y = self.gelu(y)
+        # y = self.dropout(y)
+        y = self.linear1(y)
+        # y = self.dropout(y)
+        return y
+
+
+class MixerBlock(BaseModel):
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:48]: sorry, might be a misunderstanding here, only the main model class needs to be inherited from `BaseModel`, the other should just be `ivy.Module`

File: ivy_models/mlpmixer/mlpmixer.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        y = ivy.matrix_transpose(y)
+        y = self.mlp_block1(y)
+        y = ivy.matrix_transpose(y)
+        x = x + y
+        y = self.norm2(x)
+        y = x + self.mlp_block2(y)
+        return y
+
+
+class MLPMixer(BaseModel):
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:67]: only this should be `BaseModel`

File: ivy_models/mlpmixer/mlpmixer.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        if data_format == "NCHW":
+            x = ivy.permute_dims(x, (0, 3, 1, 2))
+        x = self.conv(x)
+        x = x.reshape(
+            (int(x.shape[0]), int(x.shape[1]) * int(x.shape[2]), int(x.shape[3]))
+        )
+        for i in range(self.spec.num_blocks):
+            x = self.mixer_block(x)
+        x = self.norm(x)
+        # x = self.dropout(x)
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:135]: there's some debugging code or comment I supposed? Please clean up the unnecessary ones, thanks!

File: ivy_models/mlpmixer/mlpmixer.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        inp_dim=inp_dim,
+        num_classes=num_classes,
+        num_blocks=num_blocks,
+        patch_size=patch_size,
+        hidden_dim=hidden_dim,
+        token_mlp_dim=token_mlp_dim,
+        channel_mlp_dim=channel_mlp_dim,
+        data_format=data_format,
+    )
+    if pretrained:
```
Comment (juliagsy) [inline ivy_models/mlpmixer/mlpmixer.py:167]: for optimisation purposes, please move this as the first statement or block as a model doesn't need to be created in advance if `retrained=True`

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hello shireen! the changes look great! there are some (final I'd say) changes required for clean up and optimisation, thanks a lot!

**Review by juliagsy — APPROVED** [review]
Thanks a lot for the changes! lgtm!

**Pull request merged**